llvm/lib/CheerpWriter/PartialExecuter.cpp

```diff
@@ -1701,18 +1701,20 @@ void PartialInterpreter::visitOuter(FunctionData& data, llvm::Instruction& I)
 
 			if (next)
 			{
-				// We know where execution should proceed
-				incomingBB = I.getParent();
-				getTopCallFrame().CurBB = next;
-				getTopCallFrame().CurInst = getTopCallFrame().CurBB->begin();
+				data.incrementVisitCounter(next);
 
-				// Also here we have set the proper state for the execution so we can return
-				return;
-			}
-			else
-			{
-				skip = true;
+				if (data.getVisitCounter(next) < MAX_NUMBER_OF_VISITS_PER_BB)
```

Comment:
This looks reasonable. Currently the check is only done when starting the visit from a new block, but internal visits are not accounted for. Two things

1. Move the new check inside the existing if to better preserve the code history
2. Please validate that prints is still correctly optimized away in easy cases, we might need to bump the max number of visits since we are now counting much more often